### ISSUE TITLE
Rename Monad::bind

### ISSUE THREAD

**Issue by ldionne** [ldionne/hana#9]
The name is misleading because of the existence of `std::bind`. If anyone can come up with a better name, please suggest it.

**Comment by ldionne** [comment]
Two new suggestions:

```cpp
chain(monadic_result, monadic_function)
monadic_apply(monadic_function, monadic_result)
```

**Comment by ldionne** [comment]
Note that `monadic_apply` would be consistent with the recently added `monadic_fold`. I still don't like it as much as `chain` (or `pipe`) though.

**Comment by pfultz2** [comment]
Is it possible to combine `fold` with an overload for `monadic_fold`? Or perhaps there is a way to annotate a function to signify that it returns a monadic type?

**Comment by ldionne** [comment]
You mean provide only `fold` and use a monadic variant whenever the function returns a monadic result? This is not possible right now, because we don't know the generalized type returned by functions. It _would_ be possible to annotate functions so we have that knowledge, e.g.:
```cpp
auto f = monadic_function<Monad>([](auto x) { ... });
fold(xs, state, f); // uses a monadic fold
```

Is this what you mean?

**Comment by pfultz2** [comment]
> Is this what you mean?

Yes. Also, I wonder if `Monad` could be deduced at a later stage. So essentially you could write:

```cpp
auto f = monadic_function([](auto x) { ... });
fold(xs, state, f); // uses a monadic fold
```

**Comment by ldionne** [comment]
I assume you mean that `Monad` would be deduced when we call `monadic_function(f)` on an argument. This way, it could deduce it from the return type of the wrapped function `f`. Is this what you mean?

If so, this is a brilliant idea, but unfortunately it won't work. `monadic_fold` must know the `Monad` even when the function is never called, for example in the case of an empty sequence. This is required so it can `lift` the `state` into the proper Monad.

It _would_ be possible to wrap the state in a "yet-to-be-determined" Monad, by using e.g. `lift(state)` instead of `lift<M>(state)`. I'm not exactly sure what that would buy us and what it might complicate, but I would say it is worth thinking about.

I personnally like the fact that `monadic_fold` must be written explicitly, since they really are two different things. If we can find a nice way (e.g. a proper generalization of folding and monadic folding) to make `fold` and `monadic_fold` coincide then I would be interested, but otherwise I don't see this as a defect.

**Comment by pfultz2** [comment]
> but unfortunately it won't work. monadic_fold must know the Monad even when the function is never called

Oh yea thats right.

> I personnally like the fact that monadic_fold must be written explicitly, since they really are two different things

Yes it does make sense to have two seperate functions, as well. I was just trying to think of a possible different approach.

**Comment by viboes** [comment]
An alternative to renaming to avoid syntactic or semantic conflicts with other names is, in C++, to use namespaces to name type classes. 

```
auto x = monad::bind(m, f);
auto y = functor::map(f, m);

``` 
As Haskell as no namespaces, it needs to use prefix `f`, `m`  such as in `fmap`, `mzero`, `mplus`. 

I'm not saying that bind and map are the best names, just that adding the type class give us more freedom.

**Comment by ldionne** [comment]
While it technically gives us more freedom, in practice we then have to type `monad::bind` all around or `using namespace boost::hana::monad`. I find this unattractive. Also, when you push the concept, you end up with `foldable::fold.left`, `sequence::remove_at`, `searchable::find_if` and so on, which is cumbersome. We don't have major naming conflicts in the library right now as I've managed to name things in such a way that everything fits. Of course, the names are not perfect (as reflected by this thread), but I don't have `fmap` and `mzero` and so on either. I'd rather not use namespaces.

**Comment by ldionne** [comment]
I'm about to close this. I decided to rename `bind` to `chain`. I also want to thank @pfultz2  for suggesting the `monadic_function` thing; this made me realize that I could remove template parameters in some cases by deferring the computation, which I just did with `mcompose` (now called `monadic_compose` by the way).

**Comment by ldionne** [comment]
fixed by fb518d859afd9fd4fd0e4a4b6c1fc826fd7ac1ae
